### PR TITLE
Enable returning classes from MacroAnnotations (part 3) 

### PR DESCRIPTION
**Pull request by nicolasstucki**
Enable the addition of classes from a `MacroAnnotation`:
 * Can add new `class`/`object` definitions next to the annotated definition

Special cases:
 * An annotated top-level `def`, `val`, `var`, `lazy val` can return a `class`/`object`
   definition that is owned by the package or package object.

Related PRs:
 * Follows https://github.com/lampepfl/dotty/pull/16454

### CODE REVIEW

**Comment by michelou** [comment]
@nicolasstucki Duplicate imports in file `MacroAnnotations.scala` : 
 ```scala
import dotty.tools.dotc.core.Decorators.*
import dotty.tools.dotc.core.Decorators.*
```

File: library/src/scala/quoted/Quotes.scala [hunk 1]
```diff
@@ -3638,8 +3638,69 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
        *  @note As a macro can only splice code into the point at which it is expanded, all generated symbols must be
        *        direct or indirect children of the reflection context's owner.
        */
+      // TODO: add flags and privateWithin
       @experimental def newClass(parent: Symbol, name: String, parents: List[TypeRepr], decls: Symbol => List[Symbol], selfType: Option[TypeRepr]): Symbol
 
+      /** Generates a new module symbol with an associated module class symbol.
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:3644]:
```suggestion
      /** Generates a new module symbol with an associated module class symbol,
       *  this is equivalent to an `object` declaration in source code.
```

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 2]
```diff
@@ -2481,6 +2482,21 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
         for sym <- decls(cls) do cls.enter(sym)
         cls
 
+      def newModule(owner: Symbol, name: String, modFlags: Flags, clsFlags: Flags, parents: List[TypeRepr], decls: Symbol => List[Symbol], privateWithin: Symbol): Symbol =
+        assert(parents.nonEmpty && !parents.head.typeSymbol.is(dotc.core.Flags.Trait), "First parent must be a class")
+        val mod = dotc.core.Symbols.newCompleteModuleSymbol(
+          owner,
+          name.toTermName,
+          modFlags | Flags.Final | Flags.Lazy | Flags.Module,
+          clsFlags | Flags.Final | Flags.Module,
```
Comment (smarter) [inline compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:2491]: Could this use `ModuleValCreationFlags` and `ModuleClassCreationFlags` instead?

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 3]
```diff
@@ -2481,6 +2482,21 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
...
+          parents.asInstanceOf, // FIXME
```
Comment (smarter) [inline compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:2492]: It's better to make the type parameter of the `asInstanceOf` explicit to avoid confusion, also the FIXME should be expanded to explain what needs to be fixed.

File: compiler/src/dotty/tools/dotc/transform/MacroAnnotations.scala [hunk 4]
```diff
@@ -119,17 +119,15 @@ class MacroAnnotations(thisPhase: DenotTransformer):
...
-    else if sym.isType then
+    if sym.isType && !sym.isClass then
       report.error(i"macro annotation cannot return a `type`. $annot tried to add $sym", annot.tree)
-    else if sym.owner != annotated.owner then
+    else if sym.owner != annotated.owner && !(annotated.owner.isPackageObject && (sym.isClass || sym.is(Module)) && sym.owner == annotated.owner.owner) then
       report.error(i"macro annotation $annot added $sym with an inconsistent owner. Expected it to be owned by ${annotated.owner} but was owned by ${sym.owner}.", annot.tree)
     else if annotated.isClass && annotated.owner.is(Package) /*&& !sym.isClass*/ then
       report.error(i"macro annotation can not add top-level ${sym.showKind}. $annot tried to add $sym.", annot.tree)
-    else
+    else if !sym.is(Module) then // To avoid entering it twice
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/MacroAnnotations.scala:130]: Why is this needed?

File: library/src/scala/quoted/Quotes.scala [hunk 5]
```diff
@@ -3638,8 +3638,69 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
...
+       *    object MyModule$macro$1 extends Object:
+       *      def run(): Unit = println("run")
+       *    MyModule$macro$1.run()
+       *  }
+       *  ```
+       *
+       *  @param parent The owner of the class
+       *  @param name The name of the class
+       *  @param modFlags extra flags to with which the module symbol should be constructed
+       *  @param clsFlags extra flags to with which the module class symbol should be constructed
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:3688]: typo: "to with which"

File: library/src/scala/quoted/Quotes.scala [hunk 6]
```diff
@@ -3638,8 +3638,69 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
...
+       *
+       *  val mod = Symbol.newModule(Symbol.spliceOwner, moduleName, Flags.EmptyFlags, Flags.EmptyFlags, parents.map(_.tpe), decls, Symbol.noSymbol)
+       *  val cls = mod.moduleClass
+       *  val runSym = cls.declaredMethod("run").head
+       *
+       *  val runDef = DefDef(runSym, _ => Some('{ println("run") }.asTerm))
+       *  val clsDef = ClassDef(cls, parents, body = List(runDef))
+       *  val newCls = Apply(Select(New(TypeIdent(cls)), cls.primaryConstructor), Nil)
+       *  val modVal = ValDef(mod, Some(newCls))
+       *  val modDef = List(modVal, clsDef)
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:3666]: Having to construct both the val and class manually is a bit cumbersome, it might be nicer to have a convenience method to do both at once for this usecase.

File: library/src/scala/quoted/Quotes.scala [hunk 7]
```diff
@@ -3638,8 +3638,69 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
...
+       *    MyModule$macro$1.run()
+       *  }
+       *  ```
+       *
+       *  @param parent The owner of the class
+       *  @param name The name of the class
+       *  @param modFlags extra flags to with which the module symbol should be constructed
+       *  @param clsFlags extra flags to with which the module class symbol should be constructed
+       *  @param parents The parent classes of the class. The first parent must not be a trait.
+       *  @param decls The member declarations of the module provided the symbol of this class
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:3690]: Might be more descriptive:
```suggestion
       *  @param decls A function that takes the symbol of the module class as input and return the symbols of its declared members
```

File: library/src/scala/quoted/Quotes.scala [hunk 8]
```diff
@@ -3638,8 +3638,69 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
...
+       *  @param privateWithin the symbol within which this new method symbol should be private. May be noSymbol.
+       *
+       *  This symbol starts without an accompanying definition.
+       *  It is the meta-programmer's responsibility to provide exactly one corresponding definition by passing
+       *  this symbol to the ClassDef and ValDef constructor.
+       *
+       *  @note As a macro can only splice code into the point at which it is expanded, all generated symbols must be
+       *        direct or indirect children of the reflection context's owner.
+       *
+       *  @syntax markdown
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:3700]: Is this still needed?

File: compiler/src/dotty/tools/dotc/transform/Inlining.scala [hunk 9]
```diff
@@ -63,6 +63,12 @@ class Inlining extends MacroTransform with IdentityDenotTransformer {
   }
 
   private class InliningTreeMap extends TreeMapWithImplicits {
+
+    /** List of top level classes added by macro annotation in a package object.
+     *  These are added the PackageDef that owns this particular package object.
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/Inlining.scala:68]:
```suggestion
     *  These are added to the PackageDef that owns this particular package object.
```

File: compiler/src/dotty/tools/dotc/transform/Inlining.scala [hunk 10]
```diff
@@ -63,6 +63,12 @@ class Inlining extends MacroTransform with IdentityDenotTransformer {
...
+     */
+    private val topClasses = new collection.mutable.ListBuffer[Tree]
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/Inlining.scala:70]: Unfortunately I think this isn't sufficient because package objects can be nested:
```scala
package foo {
  val x = 1
  package bar {
    val y = 2
  }
}
```
Instead, maybe the MemberDef case of transform should return a Thicket with the top-level classes, and we should add an extra case to `transform` to handle the package object module class itself, where we should also return a Thicket with the top-level classes

File: library/src/scala/quoted/Quotes.scala (hunk 8 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala:3700]: Not sure

File: compiler/src/dotty/tools/dotc/transform/Inlining.scala (hunk 10 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/Inlining.scala:70]: This use case was considered and works. I added tests for it in `tests/run-macros/annot-add-global-class`.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/Inlining.scala:70]: Note that after after post typer the tree is
```
package foo {
  package bar {
    val y = 2
  }
  val x = 1
}
```

This implies that nested classes are processed first and the buffer never overlaps and is emptied just after transforming the nested package.

File: library/src/scala/quoted/Quotes.scala (hunk 8 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala:3700]: There are 104 instances of this in the library.

File: compiler/src/dotty/tools/dotc/transform/MacroAnnotations.scala (hunk 4 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/MacroAnnotations.scala]: Actually, I don't even seem to need the `IdentityDenotTransformer` on `MacroTransform`. Nor entering the symbols manually.

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala (hunk 3 above)
Comment (nicolasstucki) [reply compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala]: Forgot to take care of this one. I added `Symbols.newNormalizedModuleSymbol` to make this work the same way as `newClass`.

File: compiler/src/dotty/tools/dotc/transform/MacroAnnotations.scala (hunk 4 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/MacroAnnotations.scala]: I removed that code

File: library/src/scala/quoted/Quotes.scala (hunk 6 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala]: Added `ClassDef.module` to do this. I placed it there because its signature is similar to `ClassDef.apply`.

Not sure if it should return a tuple `(ValDef, ClassDef)` or a `List[Definition]`.

File: library/src/scala/quoted/Quotes.scala (hunk 1 above)
Comment (smarter) [reply library/src/scala/quoted/Quotes.scala]: It looks like this change was not applied.

File: library/src/scala/quoted/Quotes.scala [hunk 11]
```diff
@@ -467,9 +467,33 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
        *                 otherwise the can be `Term` containing the `New` applied to the parameters of the extended class.
        *  @param body List of members of the class. The members must align with the members of `cls`.
        */
+      // TODO add selfOpt: Option[ValDef]?
       @experimental def apply(cls: Symbol, parents: List[Tree /* Term | TypeTree */], body: List[Statement]): ClassDef
       def copy(original: Tree)(name: String, constr: DefDef, parents: List[Tree /* Term | TypeTree */], selfOpt: Option[ValDef], body: List[Statement]): ClassDef
       def unapply(cdef: ClassDef): (String, DefDef, List[Tree /* Term | TypeTree */], Option[ValDef], List[Statement])
+
+
+      /** Create the ValDef and ClassDef of a module.
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:476]:
```suggestion
      /** Create the ValDef and ClassDef of a module (equivalent to an `object` declaration in source code).
```

File: library/src/scala/quoted/Quotes.scala [hunk 12]
```diff
@@ -467,9 +467,33 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
...
+       *  Equivalent to
+       *  ```
+       *  def module(module: Symbol, parents: List[Tree], body: List[Statement]): (ValDef, ClassDef) =
+       *    val modCls = module.moduleClass
+       *    val modClassDef = ClassDef(modCls, parents, body)
+       *    val modValDef = ValDef(module, Some(Apply(Select(New(TypeIdent(modCls)), cls.primaryConstructor), Nil)))
+       *    List(modValDef, modClassDef)
+       *  ```
+       *
+       *  @param module the module symbol (of the module lazy val)
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:487]:
```suggestion
       *  @param module the module symbol (created using `Symbol.newModule`)
```

File: compiler/src/dotty/tools/dotc/transform/Inlining.scala (hunk 10 above)
Comment (smarter) [reply compiler/src/dotty/tools/dotc/transform/Inlining.scala:69]: > This implies that nested classes are processed first and the buffer never overlaps and is emptied just after transforming the nested package.

This is subtle, so this precondition should be documented in the code (and ideally checked somewhere, in case it breaks)

**Review by smarter — CHANGES_REQUESTED** [review]
Otherwise LGTM.

File: compiler/src/dotty/tools/dotc/transform/Inlining.scala (hunk 10 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/Inlining.scala]: I found cases where this precondition does not hold. I updated the implementation to handle such cases.

**Review by smarter — APPROVED** [review]
